implement more caching & public cache api

WebFreak001 opened the pull request:
Allows to have runtime generated bodies with proper cache

For example you can cache runtime diet files with this: (with my runtime diet PR on diet-ng)

```d
import diet.html;

if (handleCacheFile(req, res, "templates/client.dt", "public"))
	return;

res.writeBody(renderStaticHTMLDietFile("templates/client.dt"), "text/html; charset=UTF-8");
```

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +418,214 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+	MD5 md5;
+	md5.put(lastModified.stdTime.nativeToLittleEndian);
+	md5.put(dirent.size.nativeToLittleEndian);
+	etag.tag = toHexString(md5.finish())[].idup;
```
This block looks like it should be better wrapped into an `ETag` constructor or builder function?

WebFreak001 replied:
I just picked md5 and lastModified + dirent size as arbitrary choice (and because the previous algorithm was similar) and I think it would be better if this was transparent to the user so it could just be changed any time without breaking any code depending on the value of the algorithm used to generate the ETag

Changing the ETag algorithm (like the PR does) means that caches will become invalid and clients will refetch the version after the server updating vibe.d

So because the algorithm is not meant for generating a value to be used in user code and it just works with FileInfo anyway, I think it's ok to just keep it in the handleCacheFile function.

wilzbach replied:
Hmm, my point was more that this could be sth. like:

```d
struct ETag {
  static  ETag fromFile(FileInfo dirent) {
     ... (code from here)
     return etag;
  }
}
```

as imho it's not the job of this function to decide which hashing method to use.

WebFreak001 replied:
I added a helper function which does the hashing & string serialization now

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +418,214 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+			res.headers["Cache-Control"] = cache_control;
+		}
+	} else if (max_age > Duration.zero) {
+		res.headers["Cache-Control"] = "max-age=" ~ to!string(max_age.total!"seconds");
```
```suggestion
		res.headers["Cache-Control"] = text("max-age=", max_age.total!"seconds"));
```

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +418,214 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		// https://tools.ietf.org/html/rfc7232#section-3.3
+		string ifModifiedSince = req.headers.get("If-Modified-Since");
+		if (ifModifiedSince.length) {
+			const check = lastModifiedString == ifModifiedSince
```
Code style: `|| ` should be on this line

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +418,214 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		return false;
+	}
+
+	size_t i = match.indexOf('"');
```
I think using `find` could be a safer choice here and will allow you to get rid of all the indices. It will return an empty range once it's out. 
Here's an example without auto-decoding:

```d
import std.algorithm, std.range, std.stdio, std.string;

void matches(string match)
{
    for (auto r = match.representation; !r.empty; r.popFront)
    {        
        auto start = r = r.find('"').drop(1);
        r = r.find('"');
        if (r.empty) return;
        auto l = cast(char[]) start[0 .. r.ptr - start.ptr];
        writeln(l);
    }
}

void main()
{
    matches(`a = ".a." b = ".b." .`);
}
```

https://run.dlang.io/is/OMwATq

wilzbach commented on web/vibe/web/rest.d:
```diff
@@ -469,23 +469,24 @@ URLRouter registerRestInterface(TImpl)(URLRouter router, TImpl instance, string
 	generateRestJSClient!I(app, settings);
 	auto hash = app.data.md5Of.toHexString.idup;
+	ETag tag;
+	tag.tag = hash;
```
Better to use a constructor/builder that takes just `app.data` as we then have one place to change/optimize the MD5 computation.

WebFreak001 replied:
this is also just an arbitrary decision using the md5 algorithm on the contents of the file in this case. I don't really know how to better abstract something which should be an opaque string without meaning because `.md5Of.toHexString.idup` is already pretty good to use if you got the data imo

Of course the user can pick between any algorithm or use crc32 or stuff like that

wilzbach replied:
Again my ideas was to add sth. like a generic `fromBytes` to `ETag` as this looks like the wrong place to decide about the hashing algorithm.

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +418,214 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		if (s.startsWith(`W/"`)) {
+			ETag ret;
+			ret.weak = true;
+			ret.tag = s[3 .. $ - 1];
```
Could be one line:  `ETag ret = {weak: true, tag: s[3 .. $ - 1];}`

WebFreak001 commented:
(accidentally pushed irrelevant changes I made to fix tests on my machines, broke on linux and windows)

wilzbach commented on web/vibe/web/rest.d:
```diff
@@ -469,23 +469,24 @@ URLRouter registerRestInterface(TImpl)(URLRouter router, TImpl instance, string
 	auto app = appender!string();
 	generateRestJSClient!I(app, settings);
 	auto hash = app.data.md5Of.toHexString.idup;
+	ETag tag;
```
```suggestion
	ETag tag = {tag: hash}
```

wilzbach commented on web/vibe/web/rest.d:
```diff
@@ -469,23 +469,24 @@ URLRouter registerRestInterface(TImpl)(URLRouter router, TImpl instance, string
+	import vibe.http.fileserver : ETag, handleCache;
+
 	auto app = appender!string();
 	generateRestJSClient!I(app, settings);
```
Unrelated, but shouldn't we cache this?
It seems pretty silly to regenerate the entire JS client on _every_ request (the interface won't change during the application's execution), so the cache just needs to check that settings didn't change.

WebFreak001 replied:
we should create that as separate issue if it's planned to change that

WebFreak001 commented:
ping @wilzbach @s-ludwig 

Tested in firefox and chromium with properly working cache headers, adjusted code to reviews and rebased with `master`. Can I get another review or merge pls?

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		res.headers["Last-Modified"] = lastModifiedString;
+	}
+
+	if (etag != ETag.init) {
```
Doesn't look like this case is ever used? The behavior is also not mentioned in the docs.

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+			res.writeVoidBody();
+			return true;
+		}
+	}
```
1.  When recipient is the origin server and If-Match is present,
       evaluate the If-Match precondition:

       *  if true, continue to step 3
      *  if false, respond 412 (Precondition Failed) unless it can be
          determined that the state-changing request has already
          succeeded (see Section 3.1)


  2.  When recipient is the origin server, If-Match is not present, and
       If-Unmodified-Since is present, evaluate the If-Unmodified-Since
       precondition:

       *  if true, continue to step 3

tl;dr: should be `If-Unmodified-Since` here.

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+			res.statusCode = HTTPStatus.preconditionFailed;
+			res.writeVoidBody();
+			return true;
+		}
```
if true, continue to step 3

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+				res.statusCode = HTTPStatus.preconditionFailed;
+			res.writeVoidBody();
+			return true;
+		}
```
if true, continue to step 5

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		if (ifModifiedSince.length) {
+			const check = lastModifiedString == ifModifiedSince ||
+				last_modified <= parseRFC822DateTimeString(ifModifiedSince);
+			if (check) {
```
When the method is **GET or HEAD**, If-None-Match is not present, and
       If-Modified-Since is present, evaluate the If-Modified-Since
       precondition:

       *  if true, continue to step 5

       *  if false, respond 304 (Not Modified)

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		}
+
+		// https://tools.ietf.org/html/rfc7232#section-3.4
+		string ifUnmodifiedSince = req.headers.get("If-Unmodified-Since");
```
step 2 according to https://tools.ietf.org/html/rfc7232#section-6

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+				res.writeVoidBody();
+				return true;
+			}
+		}
```
Might want to add a TODO: implement If-Ranges?

```
  5.  When the method is GET and both Range and If-Range are present,
       evaluate the If-Range precondition:

       *  if the validator matches and the Range specification is
          applicable to the selected representation, respond 206
          (Partial Content) [RFC7233]
```

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+
+	Standards: https://tools.ietf.org/html/rfc7232#section-2.3.2
+*/
+bool cacheMatch(string match, ETag etag, bool allow_weak)
```
Might be worthwhile to use a Flag or enum here for better readability.

WebFreak001 replied:
this would be the first and only API in vibe.d to add Flag then, should I change it to Flag anyway?

wilzbach replied:
Yeah, I would say so. Just true/false at the caller site reads pretty terrible.

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+
+	for (auto range = allBytes; !range.empty; range.popFront)
+	{
+		auto start = range = range.find('"').drop(1);
```
This is more explicit as with the current implementation you would skip over whitespace when ` "`.

```suggestion
		range.skipOver('"');
		auto start = range;
```

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+	for (auto range = allBytes; !range.empty; range.popFront)
+	{
+		auto start = range = range.find('"').drop(1);
+		range = range.find('"');
```
Does the removal of the quotes need to happen in the loop? Start/end check are only required to be done once, no?

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+}
+
+/**
+	Matches a given match expression with a specific ETag. Can allow or disallow weak ETags and supports multiple tags.
```
does it really support multiple tags? I don't see it in the code nor are there tests for this.

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +419,254 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+			break;
+
+		const check = start[0 .. &range[0] - &start[0]];
+		const checkWeak = match[0 .. &start[0] - &allBytes[0]].endsWith(`W/"`);
```
Can't you do sth. simpler like e.g. 

```d
auto range = match.representation;
bool isWeak = range.skipOver("W/");
// if weak -> must check whether allowed (weak comparison)
if (isWeak && !allow_weak) {
  return false;
```

And now checking for the etag becomes simply:

```
// etag must start/end with "
if (range.length < 2 || range[0] != '"' || range[$ - 1] != '"')
    return false;

return range[1 .. $ -1] == etag.tag;
```

Or if you need to check for multiple etags, with a loop of course.

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +420,284 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		const check = range[0 .. end];
+		range = range[end .. $];
+		range.skipOver('"');
+		range = range.stripLeft!isWhite;
```
Move these two lines after the if as they would be needlessly executed if there's a match before.

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +420,284 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+
+	auto allBytes = match.representation;
+
+	for (auto range = allBytes; !range.empty; range.popFront)
```
No popFront here anymore as it would remove a quote if there's no whitespace (test).

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +420,284 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+	assert(cacheMatch(`"xyzzy", "r2d2xxxx", "c3piozzzz"`, ETag(false, "c3piozzzz"), true));
+
+	assert(!cacheMatch(`"xyzzy", "r2d2xxxx", "c3piozzzz"`, ETag(false, ""), false));
+	assert(!cacheMatch(`"xyzzy", "r2d2xxxx", "c3piozzzz"`, ETag(false, ""), true));
```
Can we add tests with weak and multiple tags too?

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +420,284 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+		range = range.stripLeft!isWhite;
+		bool isWeak = range.skipOver("W/");
+		if (!range.skipOver('"'))
+			break;
```
You can directly return false here (same for the other breaks).

wilzbach commented on http/vibe/http/fileserver.d:
```diff
@@ -427,3 +420,284 @@ private void sendFileImpl(scope HTTPServerRequest req, scope HTTPServerResponse
+			break;
+		auto end = range.countUntil('"');
+		if (end == -1)
+			end = range.length;
```
Isn't the quote at the end required?
